core/src/main/java/org/apache/carbondata/core/indexstore/UnsafeMemoryDMStore.java

```diff
@@ -94,87 +101,110 @@ public void addIndexRow(CarbonRowSchema[] schema, DataMapRow indexRow) throws Me
     // Check whether allocated memory is sufficient or not.
     ensureSize(rowSize);
     int pointer = runningLength;
-
+    List<Object> dataPos = new ArrayList<>();
     for (int i = 0; i < schema.length; i++) {
-      addToUnsafe(schema[i], indexRow, i);
+      addToUnsafe(schema[i], indexRow, i, dataPos);
     }
+    dataPosition.put(rowCount, dataPos);
     pointers[rowCount++] = pointer;
   }
 
-  private void addToUnsafe(CarbonRowSchema schema, DataMapRow row, int index) {
+  private void addToUnsafe(CarbonRowSchema schema, DataMapRow row, int index,
+      List<Object> dataPos) {
     switch (schema.getSchemaType()) {
       case FIXED:
         DataType dataType = schema.getDataType();
         if (dataType == DataTypes.BYTE) {
           getUnsafe()
               .putByte(memoryBlock.getBaseObject(), memoryBlock.getBaseOffset() + runningLength,
                   row.getByte(index));
-          runningLength += row.getSizeInBytes(index);
+          int sizeInBytes = row.getSizeInBytes(index);
```

Comment:
Did you verify the performance with this? how much performance is improved?

Reply:
@ravipesala attached performance report in the description. please check